File: shared/tools/type_converter.py
"""
Type conversion utilities for cross-language tool integration
"""

import json
from typing import Any, Dict, List, Union, Optional
from datetime import datetime
from decimal import Decimal
import base64


class TypeConverter:
    """Type conversion utilities"""
    
    @staticmethod
    def typescript_to_python(value: Any) -> Any:
        """Convert TypeScript types to Python types"""
        if value is None:
            return None
        
        # Handle special type indicators
        if isinstance(value, dict) and '__type__' in value:
            type_name = value['__type__']
            
            if type_name == 'datetime':
                return datetime.fromisoformat(value.get('isoformat', ''))
            
            elif type_name == 'set':
                return set(value.get('items', []))
            
            elif type_name == 'bytes':
                data = value.get('data', [])
                if isinstance(data, list):
                    return bytes(data)
                elif isinstance(data, str):
                    return base64.b64decode(data)
                return bytes()
            
            elif type_name == 'Decimal':
                return Decimal(value.get('value', '0'))
        
        # Recursively handle dicts
        if isinstance(value, dict):
            return {
                key: TypeConverter.typescript_to_python(val)
                for key, val in value.items()
            }
        
        # Recursively handle lists
        if isinstance(value, list):
            return [TypeConverter.typescript_to_python(item) for item in value]
        
        return value
    
    @staticmethod
    def python_to_typescript(value: Any) -> Any:
        """Convert Python types to TypeScript types"""
        if value is None:
            return None
        
        # Handle datetime
        if isinstance(value, datetime):
            return {
                '__type__': 'datetime',
                'isoformat': value.isoformat()
            }
        
        # Handle set
        if isinstance(value, set):
            return {
                '__type__': 'set',
                'items': list(value)
            }
        
        # Handle bytes
        if isinstance(value, bytes):
            return {
                '__type__': 'bytes',
                'data': base64.b64encode(value).decode('utf-8')
            }
        
        # Handle Decimal
        if isinstance(value, Decimal):
            return {
                '__type__': 'Decimal',
                'value': str(value)
            }
        
        # Handle tuple (mark it so TS knows it was a tuple)
        if isinstance(value, tuple):
            result = list(value)
            if isinstance(result, list):
                # Add marker that this was a tuple
                setattr(result, '__type__', 'tuple')
            return result
        
        # Recursively handle dicts
        if isinstance(value, dict):
            return {
                key: TypeConverter.python_to_typescript(val)
                for key, val in value.items()
            }
        
        # Recursively handle lists
        if isinstance(value, list):
            return [TypeConverter.python_to_typescript(item) for item in value]
        
        # Handle other types that JSON can't serialize
        try:
            json.dumps(value)
            return value
        except TypeError:
            return str(value)
    
    @staticmethod
    def json_schema_to_python_type(schema: Dict[str, Any]) -> type:
        """Convert JSON Schema to Python type hint"""
        if not schema or not isinstance(schema, dict):
            return Any
        
        schema_type = schema.get('type')
        
        if schema_type == 'string':
            return str
        elif schema_type == 'number':
            return float
        elif schema_type == 'integer':
            return int
        elif schema_type == 'boolean':
            return bool
        elif schema_type == 'array':
            item_type = TypeConverter.json_schema_to_python_type(schema.get('items', {}))
            return List[item_type]
        elif schema_type == 'object':
            return Dict[str, Any]
        elif schema_type == 'null':
            return type(None)
        else:
            return Any
    
    @staticmethod
    def validate_against_schema(value: Any, schema: Dict[str, Any]) -> tuple[bool, Optional[str]]:
        """Validate a value against a JSON schema"""
        schema_type = schema.get('type')
        
        if schema_type == 'string':
            if not isinstance(value, str):
                return False, f"Expected string, got {type(value).__name__}"
            
            if 'minLength' in schema and len(value) < schema['minLength']:
                return False, f"String length {len(value)} is less than minimum {schema['minLength']}"
            
            if 'maxLength' in schema and len(value) > schema['maxLength']:
                return False, f"String length {len(value)} is greater than maximum {schema['maxLength']}"
            
            if 'pattern' in schema:
                import re
                if not re.match(schema['pattern'], value):
                    return False, f"String does not match pattern {schema['pattern']}"
            
            if 'enum' in schema and value not in schema['enum']:
                return False, f"Value {value} is not in enum {schema['enum']}"
            
            return True, None
        
        elif schema_type == 'number' or schema_type == 'integer':
            if not isinstance(value, (int, float)):
                return False, f"Expected number, got {type(value).__name__}"
            
            if schema_type == 'integer' and not isinstance(value, int):
                return False, f"Expected integer, got float"
            
            if 'minimum' in schema and value < schema['minimum']:
                return False, f"Value {value} is less than minimum {schema['minimum']}"
            
            if 'maximum' in schema and value > schema['maximum']:
                return False, f"Value {value} is greater than maximum {schema['maximum']}"
            
            return True, None
        
        elif schema_type == 'boolean':
            if not isinstance(value, bool):
                return False, f"Expected boolean, got {type(value).__name__}"
            return True, None
        
        elif schema_type == 'array':
            if not isinstance(value, list):
                return False, f"Expected array, got {type(value).__name__}"
            
            if 'minItems' in schema and len(value) < schema['minItems']:
                return False, f"Array length {len(value)} is less than minimum {schema['minItems']}"
            
            if 'maxItems' in schema and len(value) > schema['maxItems']:
                return False, f"Array length {len(value)} is greater than maximum {schema['maxItems']}"
            
            if 'items' in schema:
                for i, item in enumerate(value):
                    valid, error = TypeConverter.validate_against_schema(item, schema['items'])
                    if not valid:
                        return False, f"Array item {i}: {error}"
            
            return True, None
        
        elif schema_type == 'object':
            if not isinstance(value, dict):
                return False, f"Expected object, got {type(value).__name__}"
            
            if 'properties' in schema:
                for prop, prop_schema in schema['properties'].items():
                    if prop in value:
                        valid, error = TypeConverter.validate_against_schema(value[prop], prop_schema)
                        if not valid:
                            return False, f"Property '{prop}': {error}"
            
            if 'required' in schema:
                for required_prop in schema['required']:
                    if required_prop not in value:
                        return False, f"Missing required property '{required_prop}'"
            
            if schema.get('additionalProperties') is False:
                allowed_props = set(schema.get('properties', {}).keys())
                actual_props = set(value.keys())
                extra_props = actual_props - allowed_props
                if extra_props:
                    return False, f"Additional properties not allowed: {extra_props}"
            
            return True, None
        
        elif schema_type == 'null':
            if value is not None:
                return False, f"Expected null, got {type(value).__name__}"
            return True, None
        
        # Handle anyOf, oneOf, allOf
        if 'anyOf' in schema:
            for sub_schema in schema['anyOf']:
                valid, _ = TypeConverter.validate_against_schema(value, sub_schema)
                if valid:
                    return True, None
            return False, "Value does not match any of the schemas"
        
        if 'oneOf' in schema:
            matches = 0
            for sub_schema in schema['oneOf']:
                valid, _ = TypeConverter.validate_against_schema(value, sub_schema)
                if valid:
                    matches += 1
            if matches == 1:
                return True, None
            elif matches == 0:
                return False, "Value does not match any of the schemas"
            else:
                return False, "Value matches multiple schemas"
        
        if 'allOf' in schema:
            for sub_schema in schema['allOf']:
                valid, error = TypeConverter.validate_against_schema(value, sub_schema)
                if not valid:
                    return False, error
            return True, None
        
        # If no type specified, accept anything
        return True, None
    
    @staticmethod
    def deep_merge(target: Dict[str, Any], source: Dict[str, Any]) -> Dict[str, Any]:
        """Deep merge two dictionaries"""
        output = target.copy()
        
        for key, value in source.items():
            if key in output and isinstance(output[key], dict) and isinstance(value, dict):
                output[key] = TypeConverter.deep_merge(output[key], value)
            else:
                output[key] = value
        
        return output
    
    @staticmethod
    def serialize_for_json(value: Any) -> Any:
        """Serialize Python objects for JSON encoding"""
        if isinstance(value, datetime):
            return value.isoformat()
        elif isinstance(value, Decimal):
            return float(value)
        elif isinstance(value, set):
            return list(value)
        elif isinstance(value, bytes):
            return base64.b64encode(value).decode('utf-8')
        elif isinstance(value, dict):
            return {
                key: TypeConverter.serialize_for_json(val)
                for key, val in value.items()
            }
        elif isinstance(value, list):
            return [TypeConverter.serialize_for_json(item) for item in value]
        else:
            return value

File: shared/tools/type-converter.ts
/**
 * Type conversion utilities for cross-language tool integration
 */

import { z } from 'zod';

export class TypeConverter {
  /**
   * Convert Python types to TypeScript types
   */
  static pythonToTypeScript(value: any): any {
    if (value === null || value === undefined) {
      return value;
    }
    
    // Handle None -> null
    if (value === 'None') {
      return null;
    }
    
    // Handle True/False -> boolean
    if (value === 'True') return true;
    if (value === 'False') return false;
    
    // Handle tuples -> arrays
    if (Array.isArray(value) && value.__type__ === 'tuple') {
      delete value.__type__;
      return value;
    }
    
    // Handle sets -> arrays
    if (value.__type__ === 'set') {
      return Array.from(value.items || []);
    }
    
    // Handle datetime -> ISO string
    if (value.__type__ === 'datetime') {
      return value.isoformat || value.toString();
    }
    
    // Handle Decimal -> number
    if (value.__type__ === 'Decimal') {
      return parseFloat(value.value);
    }
    
    // Recursively handle objects
    if (typeof value === 'object' && !Array.isArray(value)) {
      const converted: any = {};
      for (const [key, val] of Object.entries(value)) {
        converted[key] = this.pythonToTypeScript(val);
      }
      return converted;
    }
    
    // Recursively handle arrays
    if (Array.isArray(value)) {
      return value.map(item => this.pythonToTypeScript(item));
    }
    
    return value;
  }
  
  /**
   * Convert TypeScript types to Python types
   */
  static typeScriptToPython(value: any): any {
    if (value === null || value === undefined) {
      return null;
    }
    
    // Handle undefined -> None
    if (value === undefined) {
      return null;
    }
    
    // Handle Date -> datetime string
    if (value instanceof Date) {
      return {
        __type__: 'datetime',
        isoformat: value.toISOString()
      };
    }
    
    // Handle Set -> set
    if (value instanceof Set) {
      return {
        __type__: 'set',
        items: Array.from(value)
      };
    }
    
    // Handle Map -> dict
    if (value instanceof Map) {
      const obj: any = {};
      for (const [key, val] of value.entries()) {
        obj[String(key)] = this.typeScriptToPython(val);
      }
      return obj;
    }
    
    // Handle BigInt -> string
    if (typeof value === 'bigint') {
      return value.toString();
    }
    
    // Handle Buffer/Uint8Array -> bytes
    if (value instanceof Buffer || value instanceof Uint8Array) {
      return {
        __type__: 'bytes',
        data: Array.from(value)
      };
    }
    
    // Recursively handle objects
    if (typeof value === 'object' && !Array.isArray(value)) {
      const converted: any = {};
      for (const [key, val] of Object.entries(value)) {
        converted[key] = this.typeScriptToPython(val);
      }
      return converted;
    }
    
    // Recursively handle arrays
    if (Array.isArray(value)) {
      return value.map(item => this.typeScriptToPython(item));
    }
    
    return value;
  }
  
  /**
   * Convert JSON Schema to Zod schema
   */
  static jsonSchemaToZod(schema: any): z.ZodSchema {
    if (!schema || typeof schema !== 'object') {
      return z.any();
    }
    
    switch (schema.type) {
      case 'string':
        let stringSchema = z.string();
        if (schema.minLength) stringSchema = stringSchema.min(schema.minLength);
        if (schema.maxLength) stringSchema = stringSchema.max(schema.maxLength);
        if (schema.pattern) stringSchema = stringSchema.regex(new RegExp(schema.pattern));
        if (schema.enum) return z.enum(schema.enum as [string, ...string[]]);
        return stringSchema;
        
      case 'number':
      case 'integer':
        let numberSchema = schema.type === 'integer' ? z.number().int() : z.number();
        if (schema.minimum !== undefined) numberSchema = numberSchema.min(schema.minimum);
        if (schema.maximum !== undefined) numberSchema = numberSchema.max(schema.maximum);
        return numberSchema;
        
      case 'boolean':
        return z.boolean();
        
      case 'array':
        const itemSchema = schema.items ? this.jsonSchemaToZod(schema.items) : z.any();
        let arraySchema = z.array(itemSchema);
        if (schema.minItems) arraySchema = arraySchema.min(schema.minItems);
        if (schema.maxItems) arraySchema = arraySchema.max(schema.maxItems);
        return arraySchema;
        
      case 'object':
        if (schema.properties) {
          const shape: Record<string, z.ZodSchema> = {};
          for (const [key, propSchema] of Object.entries(schema.properties)) {
            shape[key] = this.jsonSchemaToZod(propSchema);
          }
          
          let objectSchema = z.object(shape);
          
          // Handle required fields
          if (schema.required && Array.isArray(schema.required)) {
            // Mark non-required fields as optional
            for (const key of Object.keys(shape)) {
              if (!schema.required.includes(key)) {
                shape[key] = shape[key].optional();
              }
            }
            objectSchema = z.object(shape);
          }
          
          // Handle additionalProperties
          if (schema.additionalProperties === false) {
            objectSchema = objectSchema.strict();
          }
          
          return objectSchema;
        }
        return z.record(z.any());
        
      case 'null':
        return z.null();
        
      default:
        // Handle anyOf, oneOf, allOf
        if (schema.anyOf) {
          const schemas = schema.anyOf.map((s: any) => this.jsonSchemaToZod(s));
          return z.union(schemas as [z.ZodSchema, z.ZodSchema, ...z.ZodSchema[]]);
        }
        
        if (schema.oneOf) {
          const schemas = schema.oneOf.map((s: any) => this.jsonSchemaToZod(s));
          return z.union(schemas as [z.ZodSchema, z.ZodSchema, ...z.ZodSchema[]]);
        }
        
        if (schema.allOf) {
          // This is a simplification - proper allOf handling is complex
          const schemas = schema.allOf.map((s: any) => this.jsonSchemaToZod(s));
          return schemas[0]; // Take the first schema as approximation
        }
        
        return z.any();
    }
  }
  
  /**
   * Convert Zod schema to JSON Schema
   */
  static zodToJsonSchema(schema: z.ZodSchema): any {
    const def = schema._def;
    
    if (def.typeName === 'ZodString') {
      const jsonSchema: any = { type: 'string' };
      if (def.checks) {
        for (const check of def.checks) {
          if (check.kind === 'min') jsonSchema.minLength = check.value;
          if (check.kind === 'max') jsonSchema.maxLength = check.value;
          if (check.kind === 'regex') jsonSchema.pattern = check.regex.source;
        }
      }
      return jsonSchema;
    }
    
    if (def.typeName === 'ZodNumber') {
      const jsonSchema: any = { type: 'number' };
      if (def.checks) {
        for (const check of def.checks) {
          if (check.kind === 'min') jsonSchema.minimum = check.value;
          if (check.kind === 'max') jsonSchema.maximum = check.value;
          if (check.kind === 'int') jsonSchema.type = 'integer';
        }
      }
      return jsonSchema;
    }
    
    if (def.typeName === 'ZodBoolean') {
      return { type: 'boolean' };
    }
    
    if (def.typeName === 'ZodArray') {
      return {
        type: 'array',
        items: this.zodToJsonSchema(def.type)
      };
    }
    
    if (def.typeName === 'ZodObject') {
      const properties: any = {};
      const required: string[] = [];
      
      for (const [key, value] of Object.entries(def.shape())) {
        properties[key] = this.zodToJsonSchema(value as z.ZodSchema);
        
        // Check if field is required
        if (!(value as any).isOptional()) {
          required.push(key);
        }
      }
      
      return {
        type: 'object',
        properties,
        required: required.length > 0 ? required : undefined,
        additionalProperties: def.strict === true ? false : true
      };
    }
    
    if (def.typeName === 'ZodNull') {
      return { type: 'null' };
    }
    
    if (def.typeName === 'ZodUnion') {
      return {
        anyOf: def.options.map((option: z.ZodSchema) => this.zodToJsonSchema(option))
      };
    }
    
    if (def.typeName === 'ZodEnum') {
      return {
        type: 'string',
        enum: def.values
      };
    }
    
    if (def.typeName === 'ZodOptional') {
      const innerSchema = this.zodToJsonSchema(def.innerType);
      return {
        ...innerSchema,
        nullable: true
      };
    }
    
    // Default fallback
    return { type: 'any' };
  }
  
  /**
   * Validate value against schema with type coercion
   */
  static validateAndCoerce(value: any, schema: z.ZodSchema, fromLanguage: 'python' | 'typescript'): any {
    // Convert value based on source language
    const converted = fromLanguage === 'python' 
      ? this.pythonToTypeScript(value)
      : this.typeScriptToPython(value);
    
    // Validate against schema
    const result = schema.safeParse(converted);
    
    if (!result.success) {
      throw new Error(`Validation failed: ${result.error.message}`);
    }
    
    return result.data;
  }
  
  /**
   * Deep merge objects (useful for combining tool configs)
   */
  static deepMerge(target: any, source: any): any {
    const output = { ...target };
    
    if (isObject(target) && isObject(source)) {
      Object.keys(source).forEach(key => {
        if (isObject(source[key])) {
          if (!(key in target)) {
            Object.assign(output, { [key]: source[key] });
          } else {
            output[key] = this.deepMerge(target[key], source[key]);
          }
        } else {
          Object.assign(output, { [key]: source[key] });
        }
      });
    }
    
    return output;
  }
}

function isObject(item: any): boolean {
  return item && typeof item === 'object' && !Array.isArray(item);
}